rust/src/quic/quic.rs

```diff
@@ -0,0 +1,338 @@
+        let cyu = Cyu::generate(&header, &data.frames);
+        QuicTransaction {
+            tx_id: 0,
+            header,
+            cyu,
+            de_state: None,
+            events: std::ptr::null_mut(),
+            tx_data: AppLayerTxData::new(),
+        }
+    }
+
+    fn free(&mut self) {
+        if !self.events.is_null() {
+            core::sc_app_layer_decoder_events_free_events(&mut self.events);
+        }
+        if let Some(state) = self.de_state {
+            core::sc_detect_engine_state_free(state);
+        }
+    }
+}
+
+impl Drop for QuicTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+pub struct QuicState {
+    tx_id: u64,
+    pub version: u32,
+    transactions: Vec<QuicTransaction>,
+}
+
+impl Default for QuicState {
+    fn default() -> Self {
+        Self {
+            tx_id: 0,
+            version: 0,
+            transactions: Vec::new(),
+        }
+    }
+}
+
+impl QuicState {
+    fn new() -> Self {
+        Self::default()
+    }
+
+    // Free a transaction by ID.
+    fn free_tx(&mut self, tx_id: u64) {
+        let tx = self
+            .transactions
+            .iter()
+            .position(|ref tx| tx.tx_id == tx_id + 1);
+        if let Some(idx) = tx {
+            let _ = self.transactions.remove(idx);
+        }
+    }
+
+    fn get_tx(&mut self, tx_id: u64) -> Option<&QuicTransaction> {
+        self.transactions.iter().find(|&tx| tx.tx_id == tx_id + 1)
+    }
+
+    fn new_tx(&mut self, header: QuicHeader, data: QuicData) -> QuicTransaction {
+        let mut tx = QuicTransaction::new(header, data);
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn tx_iterator(
+        &mut self, min_tx_id: u64, state: &mut u64,
+    ) -> Option<(&QuicTransaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+
+    fn parse(&mut self, input: &[u8]) -> bool {
+        match QuicHeader::from_bytes(input, DEFAULT_DCID_LEN) {
+            Ok((rest, header)) => match QuicData::from_bytes(rest) {
+                Ok(data) => {
+                    self.version = header.version.clone().into();
+
+                    let transaction = self.new_tx(header, data);
+                    self.transactions.push(transaction);
+
+                    return true;
+                }
+                Err(e) => {
+                    SCLogDebug!("Quic Data Parsing Error: {:?}", e);
+                    return false;
+                }
+            },
+            Err(e) => {
+                SCLogDebug!("Quic Header Parsing Error: {:?}", e);
```

Comment:
Should we set an app-layer event in these error cases ?
Grep for example `MQTTEvent::InvalidQosLevel`

Reply:
In this case a header cannot be parsed and a transaction cannot be created... were you thinking of something else ? In the case of mqtt, the header can still be parsed

Reply:
> In this case a header cannot be parsed and a transaction cannot be created... were you thinking of something else ? 

I was thinking generally about error cases.
Parsing error, either for the header or something else, may be interesting...